pkg/bgpv1/manager/reconcilerv2/pod_cidr.go

```diff
@@ -0,0 +1,194 @@
+	// Calculate desired paths per address family, collapsing per-peer advertisements into per-family advertisements.
+	desiredFamilyAdverts := make(ExportPodCIDRMetadata)
+	for _, peerFamilyAdverts := range desiredPeerAdverts {
+		for family, familyAdverts := range peerFamilyAdverts {
+			if len(familyAdverts) == 0 {
+				// if no POD CIDR advertisement is present for a peer,
+				// we can skip the peer and continue with the next peer.
+				// We do not need to evaluate individual advertisement.
+				continue
+			}
+
+			agentFamily := types.ToAgentFamily(family)
+
+			// create existingPathsSet for easier existence check.
+			existingPathsSet := sets.New[string]()
+			existingPaths, ok := desiredFamilyAdverts[agentFamily]
+			if ok {
+				for _, existingPath := range existingPaths {
+					existingPathsSet.Insert(existingPath.NLRI.String())
+				}
+			}
+
+			var paths []*types.Path
+			for _, prefix := range desiredPrefixes {
+				path := types.NewPathForPrefix(prefix)
+				path.UpdateFamilyField(agentFamily)
+
+				// skip if path already exists.
+				if existingPathsSet.Has(path.NLRI.String()) {
+					continue
+				}
+				paths = append(paths, path)
+			}
+
+			desiredFamilyAdverts[types.ToAgentFamily(family)] = append(desiredFamilyAdverts[types.ToAgentFamily(family)], paths...)
+		}
+	}
+	return desiredFamilyAdverts, nil
+}
+
+func (r *ExportPodCIDRReconciler) reconcilerFamilyAdvertisements(ctx context.Context, p ReconcileParams, desiredFamilyAdverts ExportPodCIDRMetadata) error {
+	l := r.logger.WithFields(logrus.Fields{
+		types.InstanceLogField: p.DesiredConfig.Name,
+	})
+
+	runningFamilyAdverts := r.getMetadata(p.BGPInstance)
+
+	// to delete family advertisements that are not in desiredFamilyAdverts
+	for family := range runningFamilyAdverts {
+		if _, ok := desiredFamilyAdverts[family]; !ok {
+			_, err := ExportAdvertisementsReconciler(&AdvertisementsReconcilerParams{
+				Logger:                l,
+				Ctx:                   ctx,
+				Reconciler:            r.Name(),
+				Enabled:               false,
+				Instance:              p.BGPInstance,
+				InstanceName:          p.DesiredConfig.Name,
+				CurrentAdvertisements: runningFamilyAdverts[family],
+				ToAdvertise:           nil,
+			})
+			if err != nil {
+				return err
+			}
+			delete(runningFamilyAdverts, family)
+		}
+	}
+
+	// to update family advertisements that are in both runningFamilyAdverts and desiredFamilyAdverts
+	for family := range desiredFamilyAdverts {
+		runningAdverts, err := ExportAdvertisementsReconciler(&AdvertisementsReconcilerParams{
+			Logger:                l,
+			Ctx:                   ctx,
+			Reconciler:            r.Name(),
+			Enabled:               true,
+			Instance:              p.BGPInstance,
+			InstanceName:          p.DesiredConfig.Name,
+			CurrentAdvertisements: runningFamilyAdverts[family],
+			ToAdvertise:           desiredFamilyAdverts[family],
+		})
+		if err != nil {
+			return err
+		}
+		runningFamilyAdverts[family] = runningAdverts
```

Comment:
Could we do this in a single `ExportAdvertisementsReconciler` call per family?

Reply:
I see there are two cases - delete, upsert. Both are called with different parameters of ExportAdvertisementReconciler; like enabled field. 

We could make a single call, but then deleted families might have stale family field with empty paths in metadata.
`runningFamilyAdverts[family] = nil`. I wanted to remove deleted family completely from runningFamilyAdverts map.